File: python/tests/test_rectangles_mtr.py
import unittest

import cmrdesign as cmr


class MTRRectangleTests(unittest.TestCase):
    def test_bounds_match_regression_values(self):
        y = list([x / 100 for x in range(5, 100, 10)]) * 50
        bounds = cmr.variance_bounds_martinez_taboada_ramdas(
            y,
            beta_l=0.0125,
            beta_u=0.0125,
        )
        self.assertEqual(bounds["method"], "martinez_taboada_ramdas")
        self.assertEqual(bounds["n"], len(y))
        self.assertAlmostEqual(bounds["L"], 0.04690604657062841, places=12)
        self.assertAlmostEqual(bounds["U"], 0.10320241415971833, places=12)

    def test_bounds_drop_missing_by_default(self):
        y = list([x / 100 for x in range(5, 100, 10)]) * 5
        y[3] = float("nan")
        bounds = cmr.variance_bounds_martinez_taboada_ramdas(
            y,
            beta_l=0.05,
            beta_u=0.05,
        )
        self.assertEqual(bounds["n"], len(y) - 1)
        with self.assertRaisesRegex(ValueError, "na_rm=False"):
            cmr.variance_bounds_martinez_taboada_ramdas(
                y,
                beta_l=0.05,
                beta_u=0.05,
                na_rm=False,
            )

    def test_mtr_is_available_through_applied_api(self):
        y = list([x / 100 for x in range(5, 100, 10)]) * 10
        y += list([x / 100 for x in range(10, 100, 10)]) * 12
        d = [1] * 100 + [0] * 108
        fit = cmr.cmr_two_arm(y, d, alpha=0.10, method="mtr")
        self.assertEqual(fit.method, "martinez_taboada_ramdas")
        self.assertGreaterEqual(fit.U_CMR, 0)


if __name__ == "__main__":
    unittest.main()

File: r/tests/testthat/test-rectangles-mtr.R
testthat::test_that("MTR bounds match regression values", {
  y <- rep(seq(0.05, 0.95, by = 0.10), 50)
  bounds <- variance_bounds_martinez_taboada_ramdas(
    y,
    beta_l = 0.0125,
    beta_u = 0.0125
  )

  testthat::expect_equal(bounds$method, "martinez_taboada_ramdas")
  testthat::expect_equal(bounds$n, length(y))
  testthat::expect_equal(bounds$L, 0.04690604657062841, tolerance = 1e-12)
  testthat::expect_equal(bounds$U, 0.10320241415971833, tolerance = 1e-12)
})

testthat::test_that("MTR is available through the applied API", {
  y <- c(rep(seq(0.05, 0.95, by = 0.10), 10),
         rep(seq(0.10, 0.90, by = 0.10), 12))
  d <- c(rep(1, 100), rep(0, 108))
  fit <- cmr_two_arm(y, d, alpha = 0.10, method = "mtr")

  testthat::expect_s3_class(fit, "cmr_two_arm")
  testthat::expect_equal(fit$method, "martinez_taboada_ramdas")
  testthat::expect_true(fit$U_CMR >= 0)
})

